test/gtest/ucp/test_ucp_sockaddr.cc

```diff
@@ -362,47 +411,125 @@ class test_ucp_sockaddr : public ucp_test {
+    }
+
+    void send_recv(entity& from, entity& to, send_recv_type_t send_recv_type,
+                   bool wakeup, ucp_test_base::entity::listen_cb_type_t cb_type)
+    {
+        const uint64_t send_data = ucs_generate_uuid(0);
+        void *send_req           = send(from, &send_data, sizeof(send_data),
+                                        send_recv_type, scomplete_cbx, NULL);
+        ucs_status_t send_status;
+
         uint64_t recv_data = 0;
         void *recv_req;
         if (send_recv_type == SEND_RECV_TAG) {
-            recv_req = ucp_tag_recv_nb(to.worker(), &recv_data, 1,
-                                       ucp_dt_make_contig(sizeof(recv_data)),
-                                       1, 0, rtag_complete_cb);
+            recv_req = recv(to, &recv_data, sizeof(recv_data),
+                            rtag_complete_cbx, NULL);
+        } else if (send_recv_type == SEND_RECV_STREAM) {
+            recv_req = recv(to, &recv_data, sizeof(recv_data),
+                            rstream_complete_cbx, NULL);
         } else {
-            ASSERT_TRUE(send_recv_type == SEND_RECV_STREAM);
-            ucp_stream_poll_ep_t poll_eps;
-            ssize_t              ep_count;
-            size_t               recv_length;
-            do {
-                progress();
-                ep_count = ucp_stream_worker_poll(to.worker(), &poll_eps, 1, 0);
-            } while (ep_count == 0);
-            ASSERT_EQ(1,       ep_count);
-            EXPECT_EQ(to.ep(), poll_eps.ep);
-            EXPECT_EQ(&to,     poll_eps.user_data);
-
-            recv_req = ucp_stream_recv_nb(to.ep(), &recv_data, 1,
-                                          ucp_dt_make_contig(sizeof(recv_data)),
-                                          rstream_complete_cb, &recv_length,
-                                          UCP_STREAM_RECV_FLAG_WAITALL);
+            UCS_TEST_ABORT("unsupported communication type "
+                    << send_recv_type);
         }
 
-        if (recv_req != NULL) {
-            ASSERT_TRUE(UCS_PTR_IS_PTR(recv_req));
-            while (!ucp_request_is_completed(recv_req)) {
-                check_events(from.worker(), to.worker(), wakeup, recv_req);
-            }
-            ucp_request_free(recv_req);
+        send_status = wait_req_comp(from, to, wakeup, send_req);
+        if (!check_send_status(send_status, to, recv_req, cb_type)) {
+            return;
         }
 
+        wait_req_comp(from, to, wakeup, recv_req);
         EXPECT_EQ(send_data, recv_data);
     }
 
+    void wait_for_req(void *req, entity& e,
+                      double timeout = 10.0)
+    {
+        ucs_status_t status;
+        ucs_time_t loop_end_limit =
+                ucs_get_time() + ucs_time_from_sec(timeout);
+        do {
+            ucp_worker_progress(e.worker());
+            status = ucp_request_check_status(req);
+        } while ((status == UCS_INPROGRESS) &&
+                 (ucs_get_time() < loop_end_limit));
```

Comment:
what if time is out?

Reply:
the caller must take care of it.
we just waiting for 0.5s in one case, we expect that it will be still `status == UCS_INPROGRESS`

Reply:
then maybe rename the func to `request_progress` since it does not guarantee its completion

Reply:
> then maybe rename the func to `request_progress` since it does not guarantee its completion

done